### PR TITLE
Support specifying the FT sensors to use in the estimation via multipleAnalogSensorsNames::SixAxisForceTorqueSensorsNames without the need to modify the loaded URDF

### PR DESCRIPTION
**Pull request by traversaro**
In particular:
* As also before the PR, the FT sensors to load are specified by the `multipleAnalogSensorsNames::SixAxisForceTorqueSensorsNames` parameter. This PR  just  changes the logic so that the list of FT sensors loaded by the model are modified to only include the sensors specified in the `multipleAnalogSensorsNames::SixAxisForceTorqueSensorsNames`. So, after this PR you can just specify the FT sensors to use by editing the `multipleAnalogSensorsNames::SixAxisForceTorqueSensorsNames` list, without any need to modify the URDF (the only thing that you need to make sure is that the FT sensors specified in the URDF are a superset of the one specified in `multipleAnalogSensorsNames::SixAxisForceTorqueSensorsNames` .
* The logic for the temperature compensation have been changed. If the temperature compensation is enabled, it is compulsory that all the FT sensors used provide temperature readings.
* The documentation for temperature-compensation and related parameter has been consolidated in the README.
* The `multipleAnalogSensorsNames::TemperatureSensorsNames` parameter to specify the temperature sensors to use is not supported anymore, as if the temperature compensation is used, it will be assumed that all the sensors specified in `multipleAnalogSensorsNames::SixAxisForceTorqueSensorsNames` also provide a temperature sensor with the same name.

Fix https://github.com/robotology/whole-body-estimators/issues/183 .
Fix https://github.com/robotology/whole-body-estimators/issues/32 .

### CODE REVIEW

**Comment by GiulioRomualdi** [comment]
@traversaro, this feature is a dream! I appreciate it!

**Comment by isorrentino** [comment]
Thanks a lot @traversaro! Just a curiosity, have you already tested the PR on the robot?

**Comment by traversaro** [comment]
> Thanks a lot @traversaro! Just a curiosity, have you already tested the PR on the robot?

No, I just tested it on the test added in https://github.com/robotology/whole-body-estimators/pull/177 (catching a couple of bugs). The PR does not need any change in the configuration files, so it should be quite straightforward to test on the robot.

File: devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp [hunk 1]
```diff
@@ -864,30 +897,43 @@ bool WholeBodyDynamicsDevice::openMultipleAnalogSensorRemapper(os::Searchable &c
...
-        yarp::os::Bottle & MASnamesList= MASnames.addList();
-        for(auto i=0u; i < mas_type->get(1).asList()->size(); i++)
-        {
-            MASnamesList.addString(mas_type->get(1).asList()->get(i).asString());
-        }
-        propMASRemapper.put(mas_type->get(0).asString(),MASnames.get(0));
+    // We get the multipleAnalogSensorsNames group, but from it we just use the multipleAnalogSensorsNames parameter
+    if (!prop.check("multipleAnalogSensorsNames"))
+    {
+        yError() << "wholeBodyDynamics :Impossible to find requied group multipleAnalogSensorsNames";
```
Comment (S-Dafarra) [inline devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:906]:
```suggestion
        yError() << "wholeBodyDynamics: Impossible to find required group multipleAnalogSensorsNames";
```

File: devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp [hunk 2]
```diff
@@ -864,30 +897,43 @@ bool WholeBodyDynamicsDevice::openMultipleAnalogSensorRemapper(os::Searchable &c
...
+        return false;
+    }
+
+    yarp::os::Searchable& multipleAnalogSensorsNames = prop.findGroup("multipleAnalogSensorsNames");
+
+    bool required = true;
+    bool ok = getConfigParamsAsList(multipleAnalogSensorsNames, "SixAxisForceTorqueSensorsNames", this->ftSensorNames, required);
+    if (!ok)
+    {
+        yError() << "wholeBodyDynamics :Impossible to find required variable SixAxisForceTorqueSensorsNames in group multipleAnalogSensorsNames";
```
Comment (S-Dafarra) [inline devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:916]:
```suggestion
        yError() << "wholeBodyDynamics: Impossible to find required variable SixAxisForceTorqueSensorsNames in group multipleAnalogSensorsNames";
```

File: devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp [hunk 3]
```diff
@@ -864,30 +897,43 @@ bool WholeBodyDynamicsDevice::openMultipleAnalogSensorRemapper(os::Searchable &c
...
+        return false;
+    }
 
+    if (multipleAnalogSensorsNames.find("TemperatureSensorsNames").isList())
+    {
+        yWarning() << "wholeBodyDynamics : TemperatureSensorsNames parameter is not used anymore, the names in SixAxisForceTorqueSensorsNames are used, please remove TemperatureSensorsNames parameter to avoid confusion";
```
Comment (S-Dafarra) [inline devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:922]:
```suggestion
        yWarning() << "wholeBodyDynamics: TemperatureSensorsNames parameter is not used anymore, the names in SixAxisForceTorqueSensorsNames are used, please remove TemperatureSensorsNames parameter to avoid confusion";
```

File: devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp [hunk 4]
```diff
@@ -1137,8 +1183,8 @@ bool WholeBodyDynamicsDevice::loadSettingsFromConfig(os::Searchable& config)
     std::string estimateJointVelocityAccelerationOptionName = "estimateJointVelocityAcceleration";
     if( !(prop.check(estimateJointVelocityAccelerationOptionName.c_str()) && prop.find(estimateJointVelocityAccelerationOptionName.c_str()).isBool()) )
     {
-        yWarning() << "wholeBodyDynamics: estimateVelocityAccelerationOptionName bool parameter missing, please specify it.";
-        yWarning() << "wholeBodyDynamics: setting estimateVelocityAccelerationOptionName to the default value of true, but this is a deprecated behaviour that will be removed in the future.";
+        yWarning() << "wholeBodyDynamics: estimateJointVelocityAcceleration bool parameter missing, please specify it.";
+        yWarning() << "wholeBodyDynamics: setting estimateJointVelocityAcceleration to the default value of true, but this is a deprecated behaviour that will be removed in the future.";
```
Comment (S-Dafarra) [inline devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:1187]:
```suggestion
        yWarning() << "wholeBodyDynamics: setting estimateJointVelocityAcceleration to the default value of false, but this is a deprecated behavior that will be removed in the future.";
```

File: devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp [hunk 5]
```diff
@@ -1769,140 +1828,30 @@ bool WholeBodyDynamicsDevice::attachAllVirtualAnalogSensor(const PolyDriverList&
...
-        double tic = yarp::os::Time::now();
-        bool timeSpentTryngToReadSensors = 0.0;
+        double timeSpentTryngToReadSensors = 0.0;
         bool readSuccessfull = false;
+        double tic = yarp::os::Time::now();
         while( (timeSpentTryngToReadSensors < wholeBodyDynamics_sensorTimeoutInSeconds) && !readSuccessfull )
         {
             readSuccessfull = readFTSensors(verbose);
             timeSpentTryngToReadSensors = (yarp::os::Time::now() - tic);
+            yarp::os::Time::delay(0.001);
```
Comment (S-Dafarra) [inline devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:1854]: Why is this needed?

File: devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp (hunk 4 above)
Comment (traversaro) [reply devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:1187]: Ah cool, so there is no problem in https://github.com/robotology/robots-configuration/issues/668 .

**Comment by traversaro** [comment]
The pixi crash on Windows was a byproduct of switching from Visual Studio to Ninja, by doing so the default build type switched to Debug, and conda-forge binaries are not compatible with Debug. Changing to have explicit build type for all projects fixed the problem.

File: devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp (hunk 5 above)
Comment (traversaro) [reply devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:1854]: Due to a bug, I actually ended up having a sensor check that was failing, and I realized that this part of code was extremely bugged (we were storing time in a bool : ) ). One bug that if the sensor read failed, we continued to try to read from the sensor without any kind of wait, using ~100% of the CPU. I added the delay to fix this problem. However, looking at the code now, it does not make sense to wait 1 ms if the sensor read was successful, let me change that.

Comment (traversaro) [reply devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp]: Fixed in https://github.com/robotology/whole-body-estimators/pull/194/commits/a67d28734036156f0729c537771a20338cc38803 .

**Comment by traversaro** [comment]
CI is happy, PR is ready for review/test on robot/merge.

**Review by S-Dafarra — APPROVED** [review]

**Review by GiulioRomualdi — APPROVED** [review]

**Comment by S-Dafarra** [comment]
With @LoreMoretti we tested it on the robot. We did the following modification
```diff
diff --git a/ergoCubSN002/estimators/wholebodydynamics.xml b/ergoCubSN002/estimators/wholebodydynamics.xml
index b662ac180..84e479e7b 100644
--- a/ergoCubSN002/estimators/wholebodydynamics.xml
+++ b/ergoCubSN002/estimators/wholebodydynamics.xml
@@ -27,7 +27,7 @@
   </group>
 
   <group name="multipleAnalogSensorsNames">
-    <param name="SixAxisForceTorqueSensorsNames">(l_leg_ft, l_foot_front_ft, l_foot_rear_ft, r_leg_ft, r_foot_front_ft, r_foot_rear_ft, r_arm_ft, l_arm_ft)</param>
+    <param name="SixAxisForceTorqueSensorsNames">(l_foot_front_ft, l_foot_rear_ft, r_foot_front_ft, r_foot_rear_ft, r_arm_ft, l_arm_ft)</param>
   </group>
 
```

Here is the result with the robot on the ground 
Screenshot from 2024-07-12 09-59-17

There is no force on the legs :partying_face:

**Comment by traversaro** [comment]
Great, thanks!
